### PR TITLE
[FLINK-21138][queryableState] - User ClassLoader in KvStateServerHandler

### PR DESCRIPTION
**Pull request by mproch**
## What is the purpose of the change

User ClassLoader is set as context classloader in KvStateServerHandler, so that Kryo custom serializers can be duplicated correctly.

## Brief change log

User ClassLoader is set as context classloader in KvStateServerHandler, so that Kryo custom serializers can be duplicated correctly.

## Verifying this change

Integration test is added, using custom, compiled Kryo serializer

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
   - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/itcases/AbstractQueryableStateTestBase.java [hunk 1]
```diff
@@ -361,6 +372,65 @@ public Integer getKey(Tuple2<Integer, Long> value) {
...
+        // Very important, because cluster is shared between tests and we
+        // don't explicitly check that all slots are available before
+        // submitting.
+        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(Integer.MAX_VALUE, 1000L));
+
+        // Custom serializer is not needed, it's used just to check if serialization works.
+        @SuppressWarnings("unchecked")
+        Class<Serializer<?>> customSerializerClass =
+                (Class<Serializer<?>>) userClassLoader.loadClass(customSerializerClassName);
+        env.getConfig().addDefaultKryoSerializer(Byte.class, customSerializerClass);
```
Comment (tillrohrmann) [inline flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/itcases/AbstractQueryableStateTestBase.java:398]: I think we have to register a serializer instance to make this test work. If we only use a `Class`, then the system won't try to duplicate it.

**Review by tillrohrmann — APPROVED** [review]
Thanks a lot for fixing this issue @mproch. The changes look really good to me. 

I had one comment concerning the test. I think we need to register a serializer instance instead of a `Class` in order to make the test work. I've pushed a commit which does this. Once AZP gives green light, I will merge this PR.

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/itcases/AbstractQueryableStateTestBase.java (hunk 1 above)
Comment (mproch) [reply flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/itcases/AbstractQueryableStateTestBase.java]: You're right, thanks for spotting this, I think I made one cleanup too much :)

**Comment by mproch** [comment]
Thanks @tillrohrmann. Would it be possible to cherry-pick it to 1.11 branch?

**Comment by mproch** [comment]
I think there's some problem with the test... It works when I run single test method, but fails when whole class is tested... I have to dig into that, sorry for trouble...

**Comment by tillrohrmann** [comment]
I think the problem is on the client side. In the `QueryableStateClient.createState` method we duplicate the serializers from the `StateDescriptor`. When running all tests together the thread running this operation does not have the context class loader set.

This is one of the problems with context class loaders and opaque/unclear threading models in the `QueryableStateClient` I believe.

**Comment by mproch** [comment]
You're right, that was the problem. I added explicit setting of user ClassLoader in QueryableStateClient and then pass it as context classloader in callback.

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR. LGTM. Merging it now.

**Comment by tillrohrmann** [comment]
Manually merged via aeeb6171adcd7e305e8c0c64a13ade64170fa799

**Pull request closed without merging**